### DISCUSSION TITLE
error about running tests

### DISCUSSION THREAD

**Discussion by zsh151** [neams-th-coe/cardinal#469] [Q&A] [answered]
Hi all,
I was intended to running tests after install. And then
```
en@ubuntu:~/cardinal$ ./run_tests -j8
Error! Could not find 'libmesh-config' in any of the usual libmesh's locations!
```
Is there something wrong with libmesh ?
PS:I use the base environment and already set `LIBMESH_DIR=/home/en/cardinal/contrib/moose/libmesh`

Thanks

Zhang

**Comment by zsh151** [comment]


**Reply by aprilnovak** [reply]
Can you try setting `export MOOSE_DIR=/home/en/cardinal/contrib/moose` as well?

**Reply by zsh151** [reply]
Thank you, tests could run after setting `MOOSE_DIR`.
But here comes the new problems. Some tests fail for no apparent reason. 

It seems all about NekRS. such as`FAILED(CRASH)`、`FAILED(TIMEOUT)`、`FAILED(ERRMSG)`
`FAILED(TIMEOUT)`This is because I have only assigned 4 cpus to the system and the minimum requirement for testing is 8.
What does the other two errors mean?

Here is the output.
[tests.txt](https://github.com/neams-th-coe/cardinal/files/9651084/tests.txt)

**Reply by aprilnovak** [reply]
- `FAILED(TIMEOUT)` means that the test ran for > 300 seconds, which is MOOSE's default timeout limit. The first time that you run the test suite, NekRS has to compile each case on-the-fly. This writes some build files into a `.cache/` within each test folder. If you run the test suite again, you will re-use those existing files, so it should be much faster.
- `FAILED(CRASH)` means that the test aborted. On some systems, NekRS can randomly fail to pre-compile (for instance, see this [old issue](https://github.com/Nek5000/nekRS/issues/166)). This is unfortunately outside our control, but will usually go away if you just try re-running. If the same test is persistently failing, we should report it to the NekRS team so they can investigate.
- `FAILED(ERRMSG)` means that NekRS didn't throw the error that the test was hoping to "catch." For example, we have tests that make sure the correct error messages are shown to the user, to help with things like catching input file mistakes. If NekRS isn't properly throwing the error, the test is probably crashing before the error is supposed to occur, likely due to some pre-compilation issue.

Could you please re-run the test suite and posting your `tests.txt` output again? Please save the entire output (let the suite run to completion) so that we can get a tally of how many total tests passed/failed.

**Reply by zsh151** [reply]
Thank you for your patience.

This is the entire output after re-running.
```
Ran 341 tests in 1217.7 seconds. Average test time 8.0 seconds, maximum test time 196.8 seconds.
334 passed, 25 skipped, 0 pending, 7 FAILED
```
[output.txt](https://github.com/neams-th-coe/cardinal/files/9652271/output.txt)

**Reply by aprilnovak** [reply]
That's great, looks like a lot of the timeouts are resolved now. I think the remaining failures are due to random NekRS failures to pre-compile. For instance, any error that mentions an "undefined symbol" like this is a random NekRS failure:

```
tutorials/subchannel.postprocessing: /home/en/cardinal/tutorials/subchannel/.cache/nek5000/libsfr_7pin.so: undefined symbol: updmsys_
tutorials/subchannel.postprocessing: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

Errors like this one are also random failures:

```
test:nek_file_output/avg_plugin.avg_plugin:     Message  : Error loading binary [3d69e2cd7b2d5281/binary] with dlopen: /home/en/cardinal/test/tests/nek_file_output/avg_plugin/.cache/occa/cache/3d69e2cd7b2d5281/binary: file too short
```

I think you are good to go. We really only see these errors when running the test suite, I think it is something about MOOSE's Python test environment being slightly incompatible with NekRS - you should never encounter these when running actual cases with `cardinal-opt` directly.

**Reply by zsh151** [reply]
That's wonderful!
Sincerely appreciate your help in the installation process.

**Reply by aprilnovak** [reply]
Thanks for persisting through the PETSc + libMesh builds! Sometimes they can be tricky. I'm going to mark this question as resolved, but please feel free to open a new one if anything else comes up.

**Reply by wangshuaiwu** [reply]
./run_tests -j8
When I execute this command, it says
No module named Testharness
Can you help me? Thank you very much

**Comment by aprilnovak** [comment, marked as answer]
Marking as answered via discussion above.
